CoreWidgetProvider/Helpers/ChartHelper.cs

```diff
@@ -1,138 +1,244 @@
-        return stream.ToArray();
+    public static string CreateChart(List<float> chartValues, ChartType type)
+    {
+        /* // Values to use for testing when a static image is desired.
+        chartValues.Clear();
+        chartValues = new List<float>
+        {
+            10, 30, 20, 40, 30, 50, 40, 60, 50, 100,
+            10, 30, 20, 40, 30, 50, 40, 60, 50, 70,
+            0, 30, 20, 40, 30, 50, 40, 60, 50, 70,
+        }; */
+
+        var chartDoc = new XDocument();
+
+        lock (_lock)
+        {
+            // The SVG is made of three shapes:
+            // * a colored line, plotting the points on the graph
+            // * a transparent line, outlining the gradient under the graph
+            // * a grey box, outlining the entire image
+            // The SVG also contains a definition for the fill gradient.
+            var svgElement = CreateBlankSvg(ChartHeight, ChartWidth);
+
+            // Create the line that will show the points on the graph.
+            var lineElement = new XElement(PolylineElement);
+            var points = TransformPointsToLine(chartValues, out var startX, out var finalX);
+            lineElement.SetAttributeValue(PointsAttr, points.ToString());
+            lineElement.SetAttributeValue(StyleAttr, GetLineStyle(type));
+
+            // Create the line that will contain the gradient fill.
+            TransformPointsToLoop(points, startX, finalX);
+            var fillElement = new XElement(PolylineElement);
+            fillElement.SetAttributeValue(PointsAttr, points.ToString());
+            fillElement.SetAttributeValue(StyleAttr, FillStyle);
+
+            // Add the gradient definition and the three shapes to the svg.
+            svgElement.Add(CreateGradientDefinition(type));
+            svgElement.Add(fillElement);
+            svgElement.Add(lineElement);
+            svgElement.Add(CreateBorderBox(ChartHeight, ChartWidth));
+
+            chartDoc.Add(svgElement);
+        }
+
+        return chartDoc.ToString();
+    }
+
+    private static XElement CreateBlankSvg(int height, int width)
+    {
+        var svgElement = new XElement(SvgElement);
+        svgElement.SetAttributeValue(HeightAttr, height);
+        svgElement.SetAttributeValue(WidthAttr, width);
+        return svgElement;
+    }
+
+    private static XElement CreateGradientDefinition(ChartType type)
+    {
+        var defsElement = new XElement(DefsElement);
+        var gradientElement = new XElement(LinearGradientElement);
+
+        // Vertical gradients are created when x1 and x2 are equal and y1 and y2 differ.
+        gradientElement.SetAttributeValue(X1Attr, "0%");
+        gradientElement.SetAttributeValue(X2Attr, "0%");
+        gradientElement.SetAttributeValue(Y1Attr, "0%");
+        gradientElement.SetAttributeValue(Y2Attr, "100%");
+        gradientElement.SetAttributeValue(IdAttr, "gradientId");
+
+        string stop1Style;
+        string stop2Style;
+        switch (type)

```

Comment:
What happens to this if a new widget is added that displays something about the system?  Does this switch need to be updated?

Reply:
Presumably if there were actually a new widget type added, we'd want to add new colors for it, and we'd need to update this file anyway.

Reply:
Coolio.